### PR TITLE
Kaydunov Stream

### PR DESCRIPTION
**Pull request by KaydunovDenis**
### Added new validation rules for product 
[SonarCloud](https://sonarcloud.io/dashboard?id=javaGuruBY_java-free-stream-5)  
[Coverage](https://sonarcloud.io/dashboard?id=javaGuruBY_java-free-stream-5) 
[Code Smells](https://sonarcloud.io/dashboard?id=javaGuruBY_java-free-stream-5)  
[Bugs](https://sonarcloud.io/dashboard?id=javaGuruBY_java-free-stream-5) 
[Maintainability Rating](https://sonarcloud.io/dashboard?id=javaGuruBY_java-free-stream-5)  

1. Create the following classes:  
        a. Author with fields:   
            ▪ String name  
            ▪ short age  
            ▪ List<Book> books  
        b. Book with fields  
    • String title  
    • List<Author> authors  
    • int numberOfPages  
2. Create two arrays: Author[] authors and Book[] books. 
Their elements must use elements from the neighboring array.
3. Create a stream of books and then:  
    I. check if some/all book(s) have more than 200 pages;
    II. find the books with max and min number of pages;
    III. filter books with only single author;
    IV. sort the books by number of pages/title;
    V. get list of all titles;
    VI. print them using forEach method;
    VII. get distinct list of all authors
4. Use peek method for debugging intermediate streams during execution the previous task.
5. Use parallel stream with subtask #3.
6. Analyze with mentor results of previous subtask execution, explain him the difference between stream and parallel stream.
7. Use the Optional type for determining the title of the biggest book of some author.

### CODE REVIEW

**Review by trump-ny — APPROVED** [review]

File: homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                .map(Book::getNumberOfPages)
+                .max(Integer::compareTo)
+                .orElse(0);
+    }
+
+    /**
+     * @return II. find the books with min number of pages;
+     */
+    public List<Book> findBooksWithMinCountPages() {
+        int minPages = findMinCountPagesAmongAllBooks();
```
Comment (DenisAbramovDm) [inline homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java:51]: U can use it in method without creating local variable

**Review by DenisAbramovDm — APPROVED** [review]
Seems placing test classes into general pacage is bad practice

File: homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java (hunk 1 above)
Comment (KaydunovDenis) [reply homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java:51]: How can I do this without creating a variable?
This variable is used several times.
This means that each time its value will be calculated over and over again.

Comment (KaydunovDenis) [reply homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java:51]: @DenisAbramovDm 
How can I do this without creating a variable?
This variable is used several times.
This means that each time its value will be calculated over and over again.

**Comment by KaydunovDenis** [comment]
> Seems placing test classes into general pacage is bad practice
I used standart application arhitecture.
image

File: homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java (hunk 1 above)
Comment (DenisAbramovDm) [reply homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java:51]: As I can see you are using minPages variable only once in this method

Comment (KaydunovDenis) [reply homework/kaydunov_denis/stream/src/main/java/github/kaydunovdenis/service/LibraryService.java:51]: This is not true. Inside the stream variable "minPages" is using many times during comparison
book.getNumberOfPages() == minPages

**Review by Puzzletania — APPROVED** [review]

**Review by VasiliZas — APPROVED** [review]

**Pull request merged**